frugalos_segment/src/service.rs

```diff
@@ -101,17 +101,37 @@ where
 
     fn handle_command(&mut self, command: Command) {
         match command {
-            Command::AddNode(node_id, device, client, cluster) => {
+            Command::AddNode(node_id, device, client, cluster, config) => {
                 // TODO: error handling
                 let logger = self.logger.clone();
                 let logger0 = logger.clone();
+                let logger1 = logger.clone();
+                let local_id = node_id.local_id;
                 let spawner = self.spawner.clone();
                 let rpc_service = self.rpc_service.clone();
                 let raft_service = self.raft_service.clone();
                 let mds_config = self.mds_config.clone();
                 let mds_service = self.mds_service.handle();
                 let future = device
                     .map_err(|e| track!(e))
+                    .and_then(move |device| {
+                        // raft のログが不正な状態になった場合に強制削除するための対応
+                        // https://github.com/frugalos/frugalos/issues/157
+                        // https://github.com/frugalos/raftlog/issues/18
+                        let logger = logger1.new(o!("node" => local_id.to_string()));
+                        let future = if config.discard_former_log {
+                            let storage = frugalos_raft::Storage::new(
+                                logger,
+                                local_id,
+                                device.clone(),
+                                frugalos_raft::StorageMetrics::new(),
+                            );
+                            frugalos_raft::ClearLog::new(storage)
+                        } else {
+                            frugalos_raft::ClearLog::skip()
+                        };
+                        future.map(|_| device).map_err(|e| track!(Error::from(e)))
```

Comment:
note: Here `ClearLog` is executed (conditionally), and only after that the `SegmentNode` spawns. For example, repair functionality and ClearLog cannot be run at the same time.